### PR TITLE
FSTALIGN-37: Add flag for case precision/recall, --record-case-stats

### PR DESCRIPTION
**Pull request by ajhinsvark**
Some ASR benchmarking efforts involve calculating WER while remaining case sensitive. This is an option common in a lot of other tools (e.g. https://github.com/jitsi/jiwer/blob/4fb7c4107438ff469375fbf256f8c12ee60b2884/jiwer/transforms.py#L209 ). 

Let's add this option to fstalign and not always force inputs to lower case.

### CODE REVIEW

**Comment by ajhinsvark** [comment]
Example difference in wer:
### No Case
```
TOTAL WER (96.0): 59582.0/377720.0 = 15.77%
Insertion Rate: 14998.0/377720.0 = 3.97%
Deletion Rate: 12265.0/377720.0 = 3.25%
SPEAKER SWITCH WER (96): 19179.0/81949.0 = 23.40%
┌WER Header: ITN────┬────────┬───────┬────────┐
│ Label (#files)    │ Errors │ Total │    WER │
├───────────────────┼────────┼───────┼────────┤
│ ABBREVIATION (58) │    249 │   736 │ 33.83% │
│ ALPHANUMERIC (14) │     23 │    78 │ 29.49% │
│ CARDINAL (66)     │    166 │  1528 │ 10.86% │
│ CONTRACTION (92)  │   2437 │ 17424 │ 13.99% │
│ FALLBACK (70)     │    432 │   956 │ 45.19% │
│ MONEY (20)        │      8 │   186 │   4.3% │
│ ORDINAL (19)      │     24 │    81 │ 29.63% │
│ PERCENT (29)      │     10 │   288 │  3.47% │
│ PHONE (1)         │      0 │     6 │   0.0% │
│ RANGE (4)         │      1 │    24 │  4.17% │
│ SYN_1-1 (67)      │      0 │   630 │   0.0% │
│ TIME (15)         │      4 │    49 │  8.16% │
│ WEBSITE (6)       │      5 │    29 │ 17.24% │
│ YEAR (41)         │     62 │   475 │ 13.05% │
└───────────────────┴────────┴───────┴────────┘
```
### Keep Case
```
TOTAL WER (96.0): 74012.0/377353.0 = 19.61%
Insertion Rate: 15249.0/377353.0 = 4.04%
Deletion Rate: 12149.0/377353.0 = 3.22%
SPEAKER SWITCH WER (96): 23006.0/81922.0 = 28.08%
┌WER Header: ITN────┬────────┬───────┬────────┐
│ Label (#files)    │ Errors │ Total │    WER │
├───────────────────┼────────┼───────┼────────┤
│ ABBREVIATION (58) │    318 │   513 │ 61.99% │
│ ALPHANUMERIC (14) │     40 │    72 │ 55.56% │
│ CARDINAL (66)     │    165 │  1525 │ 10.82% │
│ CONTRACTION (92)  │   3166 │ 17387 │ 18.21% │
│ FALLBACK (70)     │    484 │   858 │ 56.41% │
│ MONEY (20)        │      8 │   186 │   4.3% │
│ ORDINAL (19)      │     24 │    81 │ 29.63% │
│ PERCENT (29)      │     10 │   288 │  3.47% │
│ PHONE (1)         │      0 │     6 │   0.0% │
│ RANGE (4)         │      1 │    24 │  4.17% │
│ SYN_1-1 (44)      │      0 │   156 │   0.0% │
│ TIME (15)         │      7 │    49 │ 14.29% │
│ WEBSITE (6)       │     11 │    29 │ 37.93% │
│ YEAR (41)         │     62 │   475 │ 13.05% │
└───────────────────┴────────┴───────┴────────┘
```

### Example SBS
```
 Colorado    Colorado                                                          
  Springs    Springs 
        ......
      the    the
  Mission    mission                 ERR
      and    and
```

**Comment by ajhinsvark** [comment]
Precision/Recall Output:
```
[+++] [16:50:46] [wer] case WER, (matching words only): Precision:0.793103 Recall:0.766667
[+++] [16:50:46] [wer] case WER, (including substitutions): Precision:0.590164 Recall:0.620690
```

**Comment by ajhinsvark** [comment]
Json logging example:
```
        "caseWER" :
        {
            "precision" : 0.79310345649719238,
            "recall" : 0.76666665077209473,
            "substitutions" :
            {
                "precision" : 0.5901639461517334,
                "recall" : 0.62068963050842285
            }
        },
```

File: src/fstalign.cpp [hunk 1]
```diff
@@ -206,6 +209,21 @@ vector<shared_ptr<Stitching>> make_stitches(spWERA alignment, vector<RawCtmRecor
...
+
+      // sanity check
+      std::string token_copy = token;
+      std::transform(token_copy.begin(), token_copy.end(), token_copy.begin(), ::tolower);
+      if (hyp_tk != token_copy) {
+        logger->warn(
+            "hum, looks like the text and the alignment got out of sync? [{}] vs "
+            "[{}]",
+            hyp_tk, token_copy);
+      }
```
Comment (qmac) [inline src/fstalign.cpp:224]: Is this needed or was it for debugging purposes?

File: src/main.cpp [hunk 2]
```diff
@@ -99,6 +100,8 @@ int main(int argc, char **argv) {
                             }\n\
                         }");
 
+  get_wer->add_flag("--keep-case", keep_case, "Keep casing information, default is to only compare lowercase tokens");
```
Comment (qmac) [inline src/main.cpp:103]: Consider `--record-casing-stats` given new implementation/behavior

File: src/fstalign.cpp (hunk 1 above)
Comment (ajhinsvark) [reply src/fstalign.cpp:224]: just to match the ctm logging above

**Review by qmac — APPROVED** [review]

**Pull request merged**